File: 排序/insertSort.java
public class Main {

    // public static void main(String[] args) {
    //     int[] a={20,30,21,35,2,1};
    //     insertSortWithLookOut(a);
    //     for(int i=0;i<a.length;i++){
    //         System.out.println(a[i]);
    //     }
    // }
    // a[0]作为监视哨
    public static void insertSortWithLookOut(int[] a){
        int j;
        for(int i=2;i<a.length;i++){
            // 之前以排好序
            if(a[i]<a[i-1]){
                // 哨兵的作用
                a[0]=a[i];  
                for(j=i-1;a[j]>a[0];j--){ 
                    a[j+1]=a[j];
                }
                a[j+1]=a[0];
            }
        }
    }
}

File: 排序/insertSort.js
// 带监视哨, 在已排好的序列之后的一个元素往前扫描
const insertSortWithLookOut=(arr)=>{
  let j;
  for(let i=2;i<arr.length;i++){

    if(arr[i]<arr[i-1]){
      arr[0]=arr[i];
      for(j=i-1;arr[j]>arr[0];j--){
        arr[j+1]=arr[j];
      }
      arr[j+1]=arr[0];
    }
  }
}

// 测试: success